lib/fuelsdk/client.rb

```diff
@@ -1,242 +1,258 @@
-			end 
-		end 
-			
-		def CreateProfileAttributes(allAttributes)
-			attrs = ET_ProfileAttribute.new 
-			attrs.authStub = self
-			attrs.properties = allAttributes
-			return attrs.post
-		end
-		
-		def CreateContentAreas(arrayOfContentAreas)
-			postC = ET_ContentArea.new
-			postC.authStub = self
-			postC.properties = arrayOfContentAreas
-			sendResponse = postC.post			
-			return sendResponse
-		end
-	end
+  class Response
+    # not doing accessor so user, can't update these values from response.
+    # You will see in the code some of these
+    # items are being updated via back doors and such.
+    attr_reader :code, :message, :results, :request_id, :body, :raw
+
+    # some defaults
+    def success
+      @success ||= false
+    end
+
+    alias :success? :success
+    alias :status :success # backward compatibility
+
+    def more
+      @more ||= false
+    end
+
+    alias :more? :more
+
+    def initialize raw, client
+      @client = client # keep connection with client in case we request more
+      @results = []
+      @raw = raw
+      unpack raw
+    rescue => ex # all else fails return raw
+      puts ex.message
+      raw
+    end
+
+    def continue
+      raise NotImplementedError
+    end
+
+    private
+    def unpack raw
+      raise NotImplementedError
+    end
+  end
+
+  class Client
+    attr_accessor :debug, :access_token, :auth_token, :internal_token, :refresh_token,
+                  :id, :secret, :signature, :package_name, :package_folders, :parent_folders, :auth_token_expiration
+
+    include FuelSDK::Soap
+    include FuelSDK::Rest
+
+    def jwt= encoded_jwt
+      raise 'Require app signature to decode JWT' unless self.signature
+      decoded_jwt = JWT.decode(encoded_jwt, self.signature, true)
+
+      self.auth_token = decoded_jwt['request']['user']['oauthToken']
+      self.internal_token = decoded_jwt['request']['user']['internalOauthToken']
+      self.refresh_token = decoded_jwt['request']['user']['refreshToken']
+      self.auth_token_expiration = Time.new + decoded_jwt['request']['user']['expiresIn']
+      self.package_name = decoded_jwt['request']['application']['package']
+    end
+
+    def initialize(params={}, debug=false)
+      self.debug = debug
+      client_config = params['client']
+      if client_config
+        self.id = client_config["id"]
+        self.secret = client_config["secret"]
+        self.signature = client_config["signature"]
+      end
+
+      self.jwt = params['jwt'] if params['jwt']
+      self.refresh_token = params['refresh_token'] if params['refresh_token']
+
+      self.wsdl = params["defaultwsdl"] if params["defaultwsdl"]
+    end
+
+    def refresh force=false
+      raise 'Require Client Id and Client Secret to refresh tokens' unless (id && secret)
+      #If we don't already have a token or the token expires within 5 min(300 seconds)
+      if (self.access_token.nil? || Time.new + 300 > self.auth_token_expiration || force)
```

Comment:
not sure what you mean.... this code was already here... I agree this is not thread safe, but that is not what I am trying to solve here. I am just trying to solve the problem with the Unauthorized being raised eternally because it keeps using stale access/refresh tokens... What I mean is this won't break the gem more than it is already broken regarding the thread safety.. I think